### PR TITLE
Automatic building

### PR DESCRIPTION
**Pull request by rvdh**
This pull request will automatically build packages for fdm_materials. 

On branches, no artifacts will be saved.
If a tag is created, a package will be built with the tag as version number and pushed to the jedi-package-repository master branch.

To enable publishing to jedi-package-repository, the environment variable PACKAGE_REPO_PERSONAL_ACCESS_TOKEN needs to be set in the GitLab environment.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -26,6 +26,35 @@ endif()
 
 file(GLOB _files *.fdm_material *.sig)
 
+# Determine version number
+find_package(Git)
+
+set(GIT_VERSION "Unknown")
```
Comment (LipuFei) [inline CMakeLists.txt]: You don't need to define this. You can use `GIT_VERSION_STR` if `Git` is found. See https://cmake.org/cmake/help/v3.12/module/FindGit.html

Comment (LipuFei) [inline CMakeLists.txt]: Maybe all lowercases `unknown`? I imagine this would be better for versioning.

Comment (LipuFei) [inline CMakeLists.txt]: I confused this with the version of git you can find. I'd suggest to rename it to something like `REPO_VERSION`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -26,6 +26,35 @@ endif()
...
+if(GIT_FOUND)
+    execute_process(
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        COMMAND ${GIT_EXECUTABLE} describe --tags --exact-match --dirty
+        OUTPUT_VARIABLE GIT_VERSION
```
Comment (LipuFei) [inline CMakeLists.txt]: I'd add `RESULTS_VARIABLE` to get the exit code, and only use the output variable if the exit code is 0. See https://cmake.org/cmake/help/v3.12/command/execute_process.html

File: CMakeLists.txt [hunk 3]
```diff
@@ -26,6 +26,35 @@ endif()
...
+        ERROR_QUIET
+        OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    if(GIT_VERSION)
+        set(CPACK_PACKAGE_VERSION "${GIT_VERSION}")
+    else()
+        execute_process(
+            WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+            COMMAND ${GIT_EXECUTABLE} describe --tags --dirty
+            OUTPUT_VARIABLE GIT_VERSION
```
Comment (LipuFei) [inline CMakeLists.txt]: Same as above. Check `RESULTS_VARIABLE` before using the `OUTPUT_VARIABLE` to make it more robust.

File: CMakeLists.txt (hunk 1 above)
Comment (rvdh) [reply CMakeLists.txt]: Renamed to PACKAGE_VERSION

File: CMakeLists.txt (hunk 2 above)
Comment (rvdh) [reply CMakeLists.txt]: Used RESULT_VARIABLE instead :) Great suggestion.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -1,10 +1,38 @@
...
   - validate
+  - test
+  - build
+  - release
 
 validate:
   stage: validate
+  image: python:3.7.2-slim
   script:
     - python -m pip install lxml==4.3.0
```
Comment (ChrisTerBeke) [inline .gitlab-ci.yml]: Installing lxml takes quite some time, is there a way to cache it via a base image?

**Comment by ChrisTerBeke** [comment]
@rvdh this requires some rework now that this repo is on GitHub actions as well.

**Comment by ChrisTerBeke** [comment]
A problem I foresee with this setup is that this repo contains tags for both firmware builds and Cura builds. We don't want tags for Cura build to result in a debian package being built and published to the repo. So only tags containing 'jedi' should be published maybe?

**Comment by rvdh** [comment]
> A problem I foresee with this setup is that this repo contains tags for both firmware builds and Cura builds. We don't want tags for Cura build to result in a debian package being built and published to the repo. So only tags containing 'jedi' should be published maybe?

I've now updated the release GitHub workflow to only push on tags containing "jedi".

**Pull request merged**